README.md

```diff
@@ -76,93 +100,308 @@ us create distinct identifiers for objects in Java.
+      Person p2 = new Person("Han", 21);
+      Person p3 = new Person("Uri", 37);
+      Person p4 = p1;  //same object
+
+      System.out.println(p1.equals(p2));  //false, different objects
+      System.out.println(p1.equals(p3));  //false, different objects
+      System.out.println(p1.equals(p4));  //true, same object
+   }
 }
 ```
 
-If we were to run the code above, the result would look like this:
+As `p1` and `p4` reference the same `Person` object in memory,
+only `p1.equals(p4)` evaluates to `true`.
+
+![person equals](https://curriculum-content.s3.amazonaws.com/6677/pillars/equals_person.png)
+
+The program prints:
 
 ```text
-true
 false
+false
+true
 ```
 
 It should also be noted that when comparing `String` values that we will want
 to use the `equals()` method as the `==` only works on numbers.
 
 It is highly recommended to override the `equals()` method and the `hashCode()`
 method together when we create our own classes to ensure that the objects are
-equal in a logical sense. For example, if we had a `Bicycle` class, we would
+equal in a logical sense. Usually we consider two objects "equal" if they
+have the same state, so we override the `equals` method to compare instance
+variables. For example, if we had a `Bicycle` class, we would
 want to override the `equals()` and `hashCode()` methods to ensure that a red
-bike would be indeed equal to another red bike.
+bike would be indeed equal to another red bike, even though they are different objects.
+
+The logic for overriding `equals` is a bit complex since the parameter
+type is `Object`, not `Bicycle`.  
+
+```java
+public boolean equals(Object obj)
+```
+
+We want to compare `Bicycle` objects by comparing their `color`.
+However, the method parameter may reference an object
+that may not have a `color` field since we can
+pass in *any* class instance.  For example, the code below
+shows that we can pass either a `Bicycle` object or a `Person` object
+as an argument:
+
+
+```java
+public static void main(String[] args) {
+  Bicycle bike1 = new Bicycle("red");
+  Bicycle bike2 = new Bicycle("green");
+  Bicycle bike3 = new Bicycle("red");
+  Bicycle bike4 = bike1;
+  Bicycle bike5 = new Bicycle();
+  Person person = new Person("Fred", 50);
+
+  //true when hashcodes are equal
+  System.out.println(".equals");
+  System.out.println(bike1.equals(bike2)); //false
+  System.out.println(bike1.equals(bike3)); //true
+  System.out.println(bike1.equals(bike4)); //true
+  System.out.println(bike1.equals(bike5)); //false
+  System.out.println(bike1.equals(person)); //false
+}
+```
+
+This means we need to use an operator called `instanceof` to check
+if the parameter object is a `Bicycle`, then cast the parameter object
+to `Bicycle` before getting the `color` field.  The `instanceof` operator
+also handles the case of the parameter being `null` and immediately returns `false`.
+
+```java
+@Override
+public boolean equals(Object o) {
+    if (o == this)
```

Comment:
Can we put the curly brackets on these if statements?

Reply:
fixed